pkg/retriever/retrieve.go

```diff
@@ -235,43 +252,133 @@ func runRetrievalCandidate(ctx context.Context, cfg *RetrievalConfig, client Ret
 
-		// if query is successful, then wait for priority and execute retrieval
-		done = retrieval.waitQueue.Wait(&queryCandidate{queryResponse, time.Since(phaseStartTime)})
+		queryCandidate := &queryCandidate{queryResponse, candidate, time.Since(phaseStartTime)}
+		retrieval.queue.Put(queryCandidate)
+	}
+
+	if err != nil {
+		retrieval.sendEvent(eventpublisher.Failure(phaseStartTime, eventpublisher.QueryPhase, candidate, err.Error()))
+		if ctx.Err() != nil { // cancelled, don't report the error
+			retrieval.sendResult(retrievalResult{PhaseStart: phaseStartTime, PeerID: candidate.MinerPeer.ID})
+		} else {
+			retrieval.sendResult(retrievalResult{PhaseStart: phaseStartTime, PeerID: candidate.MinerPeer.ID, Err: err})
+		}
+
+		// Reduce attempts since we won't be making a retrieval for it
+		attempts := <-retrieval.attemptsChan
+		attempts--
+		retrieval.attemptsChan <- attempts
+	}
+}
+
+func queryPhase(
+	ctx context.Context,
+	cfg *RetrievalConfig,
+	client RetrievalClient,
+	timeout time.Duration,
+	candidate types.RetrievalCandidate,
+	onConnected func(),
+) (*retrievalmarket.QueryResponse, error) {
+
+	queryCtx := ctx // separate context so we can capture cancellation vs timeout
+
+	if timeout != 0 {
+		var timeoutFunc func()
+		queryCtx, timeoutFunc = context.WithDeadline(ctx, time.Now().Add(timeout))
+		defer timeoutFunc()
+	}
+
+	queryResponse, err := client.RetrievalQueryToPeer(queryCtx, candidate.MinerPeer, candidate.RootCid, onConnected)
+	if err != nil {
+		if ctx.Err() == nil { // not cancelled, maybe timed out though
+			log.Warnf(
+				"Failed to query miner %s for %s: %v",
+				candidate.MinerPeer.ID,
+				candidate.RootCid,
+				err,
+			)
+			return nil, fmt.Errorf("%w: %v", ErrQueryFailed, err)
+		}
+		// don't register an error on cancel, this is normal on a successful retrieval on an alternative SP
+		return nil, nil
+	}
+
+	return queryResponse, nil
+}
+
+func runRetrievalPhase(
+	ctx context.Context,
+	cfg *RetrievalConfig,
+	client RetrievalClient,
+	retrieval *retrieval,
+) {
+	retrieveSem := make(chan struct{}, 1) // only allow one retriever at a time
+	success := make(chan struct{}, 1)     // success signal
+
+	phaseStartTime := time.Now()
+
+	retrieveSem <- struct{}{}
+	for {
+		// Exit routine on success
+		select {
+		case <-success:
+			return
+		default:
+		}
+
+		// check retrieval attempts, stop looping if we're out of attempts
+		attempts := <-retrieval.attemptsChan
+		if attempts <= 0 {
+			retrieval.attemptsChan <- attempts
+			break
+		}
+		retrieval.attemptsChan <- attempts
 
-		if retrieval.canSendResult() { // move on to retrieval phase
-			// only one goroutine is allowed to execute past here at a time, so retrieval.retrievalStartTime
-			// doesn't need to be protected by a mutex, but it should mark the time the first retrieval
-			// starts
-			if retrieval.retrievalStartTime.IsZero() {
-				retrieval.retrievalStartTime = time.Now()
+		// retrieval routine
+		go func() {
```

Comment:
what is this go routine doing? It appears to be unneccesary. As I can see it, the go routine starts, then immediately the calling go routine stops until it finishes. thats... a function call, not a goroutine. There's no actual parallelism

Reply:
Correct, but that's only because we've been adhering to a single retrieval at a time. If we wanted multiple retrievals at the same time, we just change the semaphore channel buffer size.

I might have gotten overzealous with the feature there, I just thought the ability to do multiple retrievals at a time would be neat. Since this is only for a single retrieval at a time, then yes, this can be simplified. I'll change this to do so, it should be fairly easy.